src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            // Scan the list for duplicate names to provide a better error if there are duplicates.
+            var names = new HashSet<string>();
+            var duplicates = new List<string>();
+            foreach (var check in healthChecks)
+            {
+                if(!names.Add(check.Name))
+                {
+                    duplicates.Add(check.Name);
+                }
+            }
+
+            if(duplicates.Count > 0)
+            {
+                throw new InvalidOperationException($"Duplicate health checks were registered with the name(s): {string.Join(", ", duplicates)}");
+            }
+
+            Checks = healthChecks.ToDictionary(c => c.Name);
+            _logger = logger;
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                foreach (var check in Checks)
+                {
+                    _logger.LogDebug("Health check '{healthCheckName}' has been registered", check.Key);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Runs all the health checks in the application and returns the aggregated status.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> which can be used to cancel the health checks</param>
+        /// <returns>
+        /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
+        /// yielding a <see cref="CompositeHealthCheckResult"/> containing the results.
+        /// </returns>
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default)
+        {
+            var results = new Dictionary<string, HealthCheckResult>(Checks.Count);
+            foreach (var pair in Checks)
+            {
+                // If the health check does things like make Database queries using EF or backend HTTP calls,
+                // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
+                using (_logger.BeginScope(new HealthCheckLogScope(pair.Key)))
+                {
+                    HealthCheckResult result;
+                    try
+                    {
+                        _logger.LogTrace("Running health check: {healthCheckName}", pair.Key);
+                        result = await pair.Value.CheckStatusAsync(cancellationToken);
+                        _logger.LogTrace("Health check '{healthCheckName}' completed with status '{healthCheckStatus}'", pair.Key, result.Status);
+                    }
+                    catch (Exception ex)
+                    {
+                        // We don't log this as an error because a health check failing shouldn't bring down the active task.
+                        _logger.LogTrace(ex, "Health check '{healthCheckName}' threw an unexpected exception", pair.Key);
```

Comment:
Debug?

Reply:
Trace is good. Debug is a pain. Again you are debugging your app and it gets 3 or 4 lines every second due to health checks, it's just annoying.  Although in saying that a lot of places health checks are actually used filtering is heavily used.

I love .net 2.0 filtering FYI ;)

Reply:
Sure, that does actually fit my general policy of Debug being useful for a human and Trace being only really useful by log analysis tools. :)